sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,398 @@
+
+    Update labels and centers (inplace), for one iteration, distributed
+    over data chunks.
+
+    Parameters
+    ----------
+    X : {float32, float64} array-like, shape (n_samples, n_features)
+        The observations to cluster.
+
+    sample_weight : {float32, float64} array-like, shape (n_samples,)
+        The weights for each observation in X.
+
+    x_squared_norms : {float32, float64} array-like, shape (n_samples,)
+        Squared L2 norm of X.
+
+    centers_old : {float32, float64} array-like, shape (n_clusters, n_features)
+        Centers before previous iteration, placeholder for the centers after
+        previous iteration.
+
+    centers_new : {float32, float64} array-like, shape (n_clusters, n_features)
+        Centers after previous iteration, placeholder for the new centers
+        computed during this iteration.
+
+    centers_squared_norms : {float32, float64} array-like, shape (n_clusters,)
+        Squared L2 norm of the centers.
+
+    weight_in_clusters : {float32, float64} array-like, shape (n_clusters,)
+        Placeholder for the sums of the weights of every observation assigned
+        to each center.
+
+    labels : int array-like, shape (n_samples,)
+        labels assignment.
+
+    center_shift : {float32, float64} array-like, shape (n_clusters,)
+        Distance between old and new centers.
+
+    n_threads : int
+        The number of threads to be used by openmp.
+
+    update_centers : bool
+        - If True, the labels and the new centers will be computed, i.e. runs
+          the E-step and the M-step of the algorithm.
+        - If False, only the labels will be computed, i.e runs the E-step of
+          the algorithm. This is useful especially when calling predict on a
+          fitted model.
+    """
+    cdef:
+        int n_samples = X.shape[0]
+        int n_features = X.shape[1]
+        int n_clusters = centers_new.shape[0]
+
+        # hard-coded number of samples per chunk. Appeared to be close to
+        # optimal in all situations.
+        int n_samples_chunk = 256 if n_samples > 256 else n_samples
+        int n_chunks = n_samples // n_samples_chunk
+        int n_samples_rem = n_samples % n_samples_chunk
+        int chunk_idx, n_samples_chunk_eff
+        int start, end
+
+        int j, k
+
+        floating[::1] centers_squared_norms = row_norms(centers_old, squared=True)
+
+        floating *centers_new_chunk
+        floating *weight_in_clusters_chunk
+        floating *pairwise_distances_chunk
+
+    # count remainder chunk in total number of chunks
+    n_chunks += n_samples != n_chunks * n_samples_chunk
+
+    if update_centers:
+        memset(&centers_new[0, 0], 0, n_clusters * n_features * sizeof(floating))
+        memset(&weight_in_clusters[0], 0, n_clusters * sizeof(floating))
+
+    with nogil, parallel(num_threads=n_threads):
+        # thread local buffers
+        centers_new_chunk = <floating*> calloc(n_clusters * n_features, sizeof(floating))
```

Comment:
I don't know if the code would be clearer if you declared `n_threads` views instead, outside of the nogil.
It would make indexing easier to follow, probably. Just a note.

Reply:
I'm probably biased but I find the indexing clearer this way :)
Otherwise you need to get the thread num to find which view you need. I find it cumbersome.